src/compute/src/render/flat_map.rs

```diff
@@ -31,60 +39,69 @@ where
-                while let Some((extension, output_diff)) = extensions.next() {
-                    table_func_output.push((extension, output_diff));
-                    table_func_output.extend((&mut extensions).take(1023));
-                    // We could consolidate `table_func_output`, but it seems unlikely to be productive.
-                    drain_through_mfp(
-                        &input_row,
-                        &mut time,
-                        &diff,
-                        &mut datums_mfp,
-                        &table_func_output,
-                        &mfp_plan,
-                        &until,
-                        &mut output_buffer,
-                    );
-                    table_func_output.clear();
-                }
-
-                output_buffer
-                    .into_iter()
-                    .map(|(data, time, diff)| match data {
-                        Ok(row) => Ok((row, time, diff)),
-                        Err(err) => Err((err, time, diff)),
-                    })
-                    .collect::<Vec<_>>()
-            }
+        let mut storage = Vec::new();
+        let stream = ok_collection.inner;
+        let (oks, errs) = stream.unary_fallible(Pipeline, "FlatMapStage", move |_, _| {
+            Box::new(move |input, ok_output, err_output| {
+                input.for_each(|cap, data| {
+                    data.swap(&mut storage);
+
+                    let mut ok_session = ConsolidateBuffer::new(ok_output, 0);
+                    let mut err_session = ConsolidateBuffer::new(err_output, 1);
+
+                    let mut datums = DatumVec::new();
+                    let mut datums_mfp = DatumVec::new();
+
+                    for (input_row, mut time, diff) in storage.drain(..) {
+                        let temp_storage = RowArena::new();
+
+                        // Buffer for extensions to `input_row`.
+                        let mut table_func_output = Vec::new();
+
+                        // Unpack datums for expression evaluation.
+                        let datums_local = datums.borrow_with(&input_row);
+                        let args = exprs
+                            .iter()
+                            .map(|e| e.eval(&datums_local, &temp_storage))
+                            .collect::<Result<Vec<_>, _>>();
+                        let args = match args {
+                            Ok(args) => args,
+                            Err(e) => {
+                                err_session.give(&cap, (e.into(), time, diff));
+                                continue;
+                            }
+                        };
+                        let mut extensions = match func.eval(&args, &temp_storage) {
+                            Ok(exts) => exts,
+                            Err(e) => {
+                                err_session.give(&cap, (e.into(), time, diff));
+                                continue;
+                            }
```

Comment:
The `continue` weirds me out. Are you up for a minor (I think) edit here where they are replaced by `and_then` calls? Something like
```rust
let mut extensions = exprs
    .iter()
    .map(|e| e.eval(&datums_local, &temp_storage))
    .collect::<Result<Vec<_>, _>>()
    .and_then(|args| func.eval(&args, &temp_storage));

match extensions {
    Ok(exts) => { drain_through_mfp(...) },
    Err(e) => { err_session.give( ... ) },
}
```

Reply:
I wonder, is it actually correct to abort on error? I'll think a bit more about it, but great point to raise.